clang/lib/DPCT/AnalysisInfo.cpp

```diff
@@ -5310,6 +5394,64 @@ DeviceFunctionInfo::DeviceFunctionInfo(size_t ParamsNum,
       FunctionName(FunctionName), IsLambda(false) {
   ParametersProps.resize(ParamsNum);
 }
+
+void DeviceFunctionInfo::collectInfoForWrapper(const FunctionDecl *FD) {
+  if (!WrapperInfoCollected) {
```

Comment:
Move to DeviceFunctionDecl::collectInfoForWrapper.
Only generate kernelwrapper for decl that has body.